system-addon/lib/HighlightsFeed.jsm

```diff
@@ -162,7 +154,8 @@ this.HighlightsFeed = class HighlightsFeed {
         this.fetchHighlights(false);
         break;
       case at.TOP_SITES_UPDATED:
-        this.fetchHighlights(false);
+        // Only broadcast on the first TOP_SITES_UPDATED call.
+        this.fetchHighlights(false || this.highlightsLastUpdated === 0);
```

Comment:
The `false ||` is useless with the other condition. But looking at this closer, wouldn't we want to broadcast the first fetch independent of which action triggered as opposed to only for TOP_SITES_UPDATED?

Reply:
How about checking `this.store.getState().Sections[Highlights].initialized` [here](https://github.com/mozilla/activity-stream/blob/master/system-addon/lib/HighlightsFeed.jsm#L122) and dispatch when `false`.
Maybe wrap everything in a `if highlights.length > 0` as highlights can be empty for fresh profiles.

Reply:
That seems reasonable to switch to broadcasting if not previously `initialized`. Also noticed that `highlightsLength` should be unused now that we don't force fetch if fewer than 9 highlights, so that can be removed.

Not entirely sure what you're suggesting for checking `length > 0`? Don't call `updateSection` if there's nothing?

Reply:
Yes, don't broadcast if there's nothing. But then it will never show the empty state as that is `initialized && !rows.length` so nvm.